src/ILOPoolSale.sol

```diff
@@ -0,0 +1,185 @@
+            PROJECT_ID
+        );
+        require(msg.sender == _project.admin, 'UA');
+        _;
+    }
+
+    modifier duringSale() {
+        require(SALE_START <= block.timestamp, 'BS');
+        require(SALE_END > block.timestamp, 'ES');
+        _;
+    }
+
+    modifier whenCancelled() {
+        require(CANCELLED, 'NOT_CANCELLED');
+        _;
+    }
+
+    modifier whenNotCancelled() {
+        require(!CANCELLED, 'CANCELLED');
+        _;
+    }
+
+    function initialize(InitParams calldata params) external override {
+        _initialize(params.poolParams);
+        require(params.poolParams.vestingConfigs.length == 1, 'VC');
+        _vestingConfig = params.poolParams.vestingConfigs[0];
+        (SALE_START, SALE_END, MIN_RAISE, MAX_RAISE) = (
+            params.saleParams.start,
+            params.saleParams.end,
+            params.saleParams.minRaise,
+            params.saleParams.maxRaise
+        );
+    }
+
+    function buy(
+        uint256 raiseAmount,
+        address recipient
+    ) external override duringSale whenNotCancelled returns (uint256 tokenId) {
+        require(raiseAmount > 0, 'ZA');
+        // check if raise amount over capacity
+        uint256 totalRaised = TOTAL_RAISED;
+        require(MAX_RAISE - totalRaised >= raiseAmount, 'MR');
+        totalRaised += raiseAmount;
+        TOTAL_RAISED = totalRaised;
+
+        // if investor already have a position, just increase raise amount and liquidity
+        // otherwise, mint new nft for investor and assign vesting schedules
+        if (balanceOf(recipient) == 0) {
+            _mint(recipient, (tokenId = _nextId++));
+            _positionVests[tokenId].schedule = _vestingConfig.schedule;
+        } else {
+            tokenId = tokenOfOwnerByIndex(recipient, 0);
+        }
+
+        Position storage _position = _positions[tokenId];
+
+        uint256 _allocation = allocation(recipient);
+        // we need to check `_position.raiseAmount < _allocation`
+        // because project admin can change whitelist during sale.
+        // so, technicaly, `_position.raiseAmount` can bigger than `_allocation`
+        // second reason is to avoid overflow of subtraction.
+        require(
+            _position.raiseAmount < _allocation &&
+                raiseAmount <= _allocation - _position.raiseAmount,
+            'UC'
+        );
+        _position.raiseAmount += raiseAmount;
+
+        // transfer fund into contract
+        TransferHelper.safeTransferFrom(
+            PAIR_TOKEN,
+            msg.sender,
+            address(this),
+            raiseAmount
+        );
+
+        emit Buy(recipient, tokenId, raiseAmount);
+    }
+
+    /// @inheritdoc IILOPoolBase
+    function launch(
+        address uniV3PoolAddress,
+        PoolAddress.PoolKey calldata poolKey,
+        uint160 sqrtPriceX96
+    ) external override OnlyManager onlyInitializedProject whenNotCancelled {
+        if (block.timestamp > SALE_END) {
```

Comment:
Just noting down 1st as im seeing many operational risk here. If project is fucked up on the timeline, they will have to redeploy everything. 

Note down 1st & come back later